float32 vr
float32 vl
bool ifUnload